src/modules/kvs/cache.c

```diff
@@ -295,12 +297,24 @@ struct cache_entry *cache_lookup (struct cache *cache, const char *ref,
     return entry;
 }
 
-void cache_insert (struct cache *cache, const char *ref,
-                   struct cache_entry *entry)
+int cache_insert (struct cache *cache, const char *ref,
+                  struct cache_entry *entry)
 {
-    int rc = zhash_insert (cache->zh, ref, entry);
-    assert (rc == 0);
-    zhash_freefn (cache->zh, ref, cache_entry_destroy);
+    int rc;
+
+    if (cache && ref && entry) {
+        /* We should not be inserting the same entry twice */
+        assert (!entry->blobref);
+
+        if (!(entry->blobref = strdup (ref)))
```

Comment:
See comment about maybe switching to zhashx_t (I made it outside of the review, sorry)